### ISSUE TITLE
How to get all users who have no role assigned?

### ISSUE THREAD

**Issue by orchardc** [OrchardCMS/OrchardCore#2987]
GetUsersInRoleAsyn("Role_Name") can only get users with the role: Role_Name
But how to get all users who have **no role assigned**?

Thanks!

**Comment by orchardc** [comment]
Also I found it is not possible to even get all site users !

When using  `_userManager.Users` threw error:  "Store does not implement IQueryableUserStore<TUser>."

**Comment by orchardc** [comment]
Thanks @sebastienros , does this mean currently we can't get all users with Orchard Core?

**Comment by Skrypt** [comment]
You can uae a Query temporarly ...

**Comment by orchardc** [comment]
@Skrypt , it would be great to see what's query can ### get users without any role assigned?

**Comment by Skrypt** [comment]
On SQLite something like this should work :

```SQL
SELECT T1.*, json_extract(T1.Content, '$.RoleNames') AS Roles FROM Document AS T1 WHERE T1.Type = 'OrchardCore.Users.Models.User, OrchardCore.Users.Core'  AND Roles == '[]'
```

**Comment by orchardc** [comment]
Thanks @Skrypt , 

I'm using vs2017 with Dapper and Dapper Contrib Extensions.

When run your SQL query, I got the error: SQL logic error **no such function: json_extract**

 Am I miss anything?

**Comment by orchardc** [comment]
PR created: https://github.com/OrchardCMS/OrchardCore/issues/2987

**Comment by Skrypt** [comment]
- Ok, first, you will go to : https://github.com/OrchardCMS/OrchardCore
- You will click on the "Fork" button on the top right corner of that page. It will add the OrchardCore repository to your own Github account as a "Fork".
- You will Clone that repository locally on your computer (with an app like GitExtensions) or make changes directly to the files on Github on your account fork of Orchard Core.
- If you go back to the main page of your fork then you should see a "create a PR" button that will display the same "compare tools" that is on the screenshot you sent. It should then display the differences between your fork branch and the main Orchard Core repo branch you want to commit these changes to "normally the dev branch".

If you can't do these steps and seem lost ; I suggest you find a tutorial about how to use Github on Google.

**Comment by sebastienros** [comment]
It's not a branch, it's a class that needs a new method on it.

**Comment by orchardc** [comment]
OK, I create a pull under my repository now:
orchardc#1

**Comment by choXberg** [comment]
do you mean something like this:

src/OrchardCore/OrchardCore.Users.Core/Services/UserStore.cs:
```
public async Task<IList<IUser>> GetAllUsers(CancellationToken cancellationToken)
{
            var users = await _session.Query<User, UserIndex>().ListAsync();
            return users == null ? new List<IUser>() : users.ToList<IUser>();
}
```
the UserIndex already exists at src/OrchardCore/OrchardCore.Users.Core/Indexes/UserIndex.cs

do we need pagination too?

**Comment by Skrypt** [comment]
Yes that's what he meant and yes I think it would be better to have pagination.

**Comment by choXberg** [comment]
getting all users without pagingation could be a performance issue for big sites, but I am not sure.
we could set default parameters and wrapped return type like PagedList<IUser>: 

PagedList<IUser> GetAllUsersPaged( CancellationToken cancellationToken, pageIndex =0, pageSize=50) 
maybe it's better to overload the function:

IUser GetAllUsers()//Clients should be aware of performance issues for huge list of users

if this issue is still unassigned, I could take it - first version without pagination

**Comment by Skrypt** [comment]
Not if the request is cached.

**Comment by sondli** [comment]
Any updates on this issue? Having the same problem with fetching all users

**Comment by ns8482e** [comment]
Orchard core’s user manager doesn’t support query all users, However you can use yessql session to query users

**Comment by sebastienros** [comment]
Since it's not available in the `UserManager` class from Microsoft Identity, I'd suggest to add it to `IUserService`.
